### ISSUE TITLE
Bug: TypeError [ERR_UNKNOWN_FILE_EXTENSION]: Unknown file extension ""

### ISSUE THREAD

**Issue by Luna-Klatzer** [oclif/oclif#712]
What is the current behavior?
-----------------------------

Currently, even with a successful build using `tsc` the `/bin/run` command fails with the following error:

```bash
node:internal/errors:464
    ErrorCaptureStackTrace(err);
    ^

TypeError [ERR_UNKNOWN_FILE_EXTENSION]: Unknown file extension "" for D:\Dokumente\GAWD\Programmierung\Web-Dev\TypeScript\Project\Kipper\kipper\bin\run
←[90m    at new NodeError (node:internal/errors:371:5)←[39m
←[90m    at Object.file: (node:internal/modules/esm/get_format:72:15)←[39m
←[90m    at defaultGetFormat (node:internal/modules/esm/get_format:85:38)←[39m
←[90m    at defaultLoad (node:internal/modules/esm/load:13:42)←[39m
←[90m    at ESMLoader.load (node:internal/modules/esm/loader:303:26)←[39m
←[90m    at ESMLoader.moduleProvider (node:internal/modules/esm/loader:230:58)←[39m
←[90m    at new ModuleJob (node:internal/modules/esm/module_job:63:26)←[39m
←[90m    at ESMLoader.getModuleJob (node:internal/modules/esm/loader:244:11)←[39m
    at async Promise.all (index 0)
←[90m    at async ESMLoader.import (node:internal/modules/esm/loader:281:24)←[39m {
  code: ←[32m'ERR_UNKNOWN_FILE_EXTENSION'←[39m
}
```

## To reproduce
- Use node `v16.13.0` on Windows `10.0.19043` Build 19043
- Use `oclif@1.18.1` with the default configuration
- Run `npx oclif multi <name>` and select TS and none of the CI extensions
- Answer yes to eslint
- Answer yes to mocha
- Run `tsc` to compile the source code
- Run `/bin/run` and view error (for me)

## `package.json`

```json
{
  "name": "kipper",
  "description": "The Kipper programming language (CLI implementation)",
  "version": "0.0.2",
  "author": "Luna-Klatzer @Luna-Klatzer",
  "bin": {
    "kipper": "./bin/run"
  },
  "bugs": "https://github.com/Luna-Klatzer/kipper/issues",
  "devDependencies": {
    "@oclif/dev-cli": "^1.26.0",
    "@oclif/test": "^1.2.8",
    "@types/chai": "^4.2.22",
    "@types/mocha": "^5.2.7",
    "@types/node": "^10.17.60",
    "chai": "^4.3.4",
    "eslint": "^7.11.0",
    "eslint-config-oclif": "^3.1.0",
    "eslint-config-oclif-typescript": "^0.1.0",
    "globby": "^10.0.2",
    "mocha": "^5.2.0",
    "nyc": "^14.1.1",
    "rimraf": "^3.0.2",
    "ts-node": "^8.10.2",
    "typescript": "^3.9.10"
  },
  "engines": {
    "node": ">=12.0.0"
  },
  "files": [
    "/bin",
    "/lib",
    "/npm-shrinkwrap.json",
    "/oclif.manifest.json"
  ],
  "homepage": "https://github.com/Luna-Klatzer/kipper",
  "keywords": [
    "oclif"
  ],
  "license": "GPL-3.0",
  "main": "lib/index.js",
  "oclif": {
    "commands": "./lib/commands",
    "bin": "kipper",
    "plugins": [
      "@oclif/plugin-help"
    ]
  },
  "repository": "Luna-Klatzer/kipper",
  "scripts": {
    "postpack": "rimraf oclif.manifest.json",
    "posttest": "npm run tslint",
    "prepack": "rimraf lib && tsc -b && oclif-dev manifest && oclif-dev readme",
    "test": "nyc --extension .ts mocha --forbid-only \"test/**/*.test.ts\"",
    "version": "oclif-dev readme && git add README.md",
    "prebuild": "npm run tslint",
    "build": "tsc",
    "build:watch": "tsc --watch",
    "rebuild": "npm run build",
    "lint": "npm run tslint && npm run prettier",
    "lint:fix": "npm run tslint:fix && npm run prettier",
    "prettier": "prettier --write ./src/",
    "tslint": "eslint src/**/*.ts --ext .ts --config .eslintrc",
    "tslint:fix": "eslint src/**/*.ts --fix --ext .ts --config .eslintrc && npm run prettier",
    "antlr4ts": "antlr4ts -visitor -o src/compiler/parser/ Kipper.g4"
  },
  "types": "lib/index.d.ts",
  "dependencies": {
    "@oclif/command": "^1.8.0",
    "@oclif/config": "^1.17.0",
    "@oclif/plugin-help": "^3.2.4",
    "tslib": "^1.14.1",
    "antlr4ts": "^0.5.0-alpha.4",
    "antlr4ts-cli": "^0.5.0-alpha.4"
  },
  "prettier": {
    "tabWidth": 2,
    "singleQuote": false
  },
  "type": "module"
}
```

I also was able to get this behaviour with the following configuration file that is in my stable project:
```json
{
  "name": "kipper",
  "description": "The Kipper programming language (CLI implementation)",
  "version": "0.0.2",
  "author": "Luna-Klatzer @Luna-Klatzer",
  "bin": {
    "kipper": "./bin/run.cjs"
  },
  "dependencies": {
    "antlr4ts": "^0.5.0-alpha.4",
    "antlr4ts-cli": "^0.5.0-alpha.4",
    "@oclif/command": "^1.8.0",
    "@oclif/config": "^1.17.0",
    "@oclif/plugin-help": "^3.2.4",
    "tslib": "^1.14.1"
  },
  "devDependencies": {
    "@oclif/dev-cli": "^1.26.0",
    "@oclif/test": "^1.2.8",
    "@types/chai": "^4.2.22",
    "@types/mocha": "^9.0.0",
    "@typescript-eslint/eslint-plugin": "^5.3.1",
    "chai": "^4.3.4",
    "eslint": "^8.2.0",
    "eslint-plugin-unicorn": "^38.0.1",
    "eslint-config-oclif": "^4.0.0",
    "eslint-config-oclif-typescript": "^1.0.2",
    "globby": "^11.0.1",
    "mocha": "^8.2.1",
    "nyc": "^15.1.0",
    "rimraf": "^3.0.2",
    "typescript": "^4.4.3",
    "prettier": "^2.4.1",
    "@sinonjs/commons": "^1.8.3",
    "@sinonjs/eslint-config": "^4.0.4",
    "@sinonjs/fake-timers": "^8.1.0",
    "@types/sinonjs__fake-timers": "~6.0.0",
    "ts-node": "^9.0.0"
  },
  "engines": {
    "node": ">=12.0.0"
  },
  "files": [
    "/bin",
    "/lib",
    "/npm-shrinkwrap.json",
    "/oclif.manifest.json"
  ],
  "homepage": "https://github.com/Luna-Klatzer/Kipper",
  "bugs": "https://github.com/Luna-Klatzer/Kipper/issues",
  "repository": {
    "type": "git",
    "url": "https://github.com/Luna-Klatzer/Kipper"
  },
  "keywords": [
    "oclif",
    "language",
    "compiler",
    "cli",
    "C"
  ],
  "license": "GPL-3.0",
  "main": "lib/index.js",
  "oclif": {
    "commands": "./lib/commands",
    "bin": "kipper",
    "plugins": [
      "@oclif/plugin-help"
    ]
  },
  "scripts": {
    "postpack": "rimraf oclif.manifest.json",
    "posttest": "npm run lint:fix",
    "prepack": "rimraf lib && tsc -b && oclif-dev manifest && oclif-dev readme",
    "test": "nyc --extension .ts mocha --forbid-only \"**/*.test.ts\"",
    "version": "oclif-dev readme && git add README.md",
    "prebuild": "npm run tslint",
    "build": "tsc",
    "build:watch": "tsc --watch",
    "rebuild": "npm run build",
    "lint": "npm run tslint && npm run prettier",
    "lint:fix": "npm run tslint:fix && npm run prettier",
    "prettier": "prettier --write ./src/",
    "tslint": "eslint src/**/*.ts || true",
    "tslint:fix": "eslint src/**/*.ts --fix || true && npm run prettier",
    "antlr4ts": "antlr4ts -visitor -o src/compiler/parser/ Kipper.g4"
  },
  "types": "lib/index.d.ts",
  "prettier": {
    "tabWidth": 2,
    "singleQuote": false
  },
  "type": "module"
}
```

What is the expected behaviour?
---------------------------------

The CLI should run without problems and execute properly.

**Comment by fed-gren** [comment]
I have the same issue. developing node using typescript, I know that need `"type": "module"` in package.json to use import/export. But when I remove `"type": "module"` it works fine. Why?

**Comment by marckraw** [comment]
Exactly, does Oclif support native esm ?

**Comment by semmerich-fliptrx** [comment]
Hi all, my company is using oclif for our internal tooling and I am running into the same issue. From what I have seen in the issues v2 should support ESM and COMMONJS. Any ideas on this?

**Comment by 0xBC84** [comment]
Any update on this? Using projects such as [lowdb](https://github.com/typicode/lowdb) completely break oclif.

**Comment by youyiqin** [comment]
> Any update on this? Using projects such as [lowdb](https://github.com/typicode/lowdb) completely break oclif.

Yes,Lowdb can not use with oclif now.Any one know how to do?Thanks!

**Comment by jercle** [comment]
Hi all,

My workaround for this is while building https://github.com/jercle/azgo/tree/rebuild-cli-using-oclif/azgoc
Just a note: The `azgoc` path is specifically rebuilding the CLI portion of the application using oclif instead of yargs.
 
Changing `#!/usr/bin/env node` to `#!/usr/bin/env node --loader ts-node/esm` in the `./bin/run` and `./bin/dev` files resolves the issue for me.

@Developer27149 - I'm yet to try this with Lowdb

**Comment by SergeyPoluektov** [comment]
Another workaround is to use [`esm`](https://www.npmjs.com/package/esm) package and require it in shebang line `#!/usr/bin/env node -r esm`. This one is taken from [`uvu`](https://github.com/lukeed/uvu/blob/master/docs/esm.md#polyfill--via-esm-package)

**Comment by baltpeter** [comment]
There is now an official documentation page on how to solve this: https://oclif.io/docs/esm

**Comment by jasonk** [comment]
After following the documentation there is still a situation where this doesn't work. If you use `plugins link` to link a plugin for development (and the plugin has commands in TypeScript), trying to run any of the commands from the plugin throws the `ERR_UNKNOWN_FILE_EXTENSION` error.

**Comment by jasonk** [comment]
For anybody else who ends up here and was having the same problem I was having where `plugins link` would still give you the `ERR_UNKNOWN_FILE_EXTENSION` error I did finally find a way to work around it. It unfortunately means that it won't transpile your plugins when running, so you'll have to make sure to run `npm run build` before running your command, but at least it will work.

To do that you can disable `ts-node` support by setting `global.oclif = { tsnodeEnabled : false };` to your `<command>-dev.js` file, *before* you import any `oclif` modules.

**Comment by mdonnalley** [comment]
This issue will be fixed in https://github.com/oclif/core/pull/759

**Comment by mdonnalley** [comment]
Fixed by https://github.com/oclif/core/pull/759 (currently available in `@ocilf/core@3.0.0-beta.5`)
